src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1041,102 +1041,184 @@ private bool IsValidFixedVariableInitializer(TypeSymbol declType, SourceLocalSym
+                            ((object)getPinnableMethod == null || !extensisbleFixedEnabled || getPinnableMethod.ContainingType.SpecialType != SpecialType.System_String))
+                        {
+                            elementType = this.GetSpecialType(SpecialType.System_Char, diagnostics, initializerSyntax);
+                            Debug.Assert(!elementType.IsManagedType);
                             break;
-                        case BoundKind.FieldAccess:
-                            var fa = (BoundFieldAccess)initializerOpt;
-                            if (!fa.FieldSymbol.IsFixed)
+                        }
+
+                        if (extensisbleFixedEnabled)
+                        {
+                            // not a specially known type, check for getPinnableMethod
+                            if (getPinnableMethod != null)
                             {
-                                Error(diagnostics, ErrorCode.ERR_FixedNotNeeded, initializerSyntax);
-                                return true;
+                                elementType = getPinnableMethod.ReturnType;
+                                break;
                             }
-                            else
+                            else if (additionalDiagnostics != null)
                             {
-                                elementType = ((PointerTypeSymbol)fa.Type).PointedAtType;
+                                diagnostics.AddRange(additionalDiagnostics);
                             }
-                            break;
-                        case BoundKind.Conversion:
-                            // The following assertion would not be correct because there might be an implicit conversion after (above) the explicit one.
-                            //Debug.Assert(((BoundConversion)initializerOpt).ExplicitCastInCode, "The assignment conversion hasn't been applied yet, so this must be from source.");
-
-                            // NOTE: Dev10 specifically doesn't report this error for the array or string cases.
-                            Error(diagnostics, ErrorCode.ERR_BadCastInFixed, initializerSyntax);
-                            return true;
-                        default:
-                            // CONSIDER: this is a very confusing error message, but it's what Dev10 reports.
-                            Error(diagnostics, ErrorCode.ERR_FixedNotNeeded, initializerSyntax);
-                            return true;
+                        }
+                        else if (getPinnableMethod != null)
+                        {
+                            CheckFeatureAvailability(initializerOpt.Syntax, MessageID.IDS_FeatureExtensibleFixedStatement, diagnostics);
+                        }
+
+                        Error(diagnostics, ErrorCode.ERR_ExprCannotBeFixed, initializerSyntax);
                     }
+                    finally
```

Comment:
This `finally` seems overly complicated. Couldn't we just stick the `Error` call in an `else` and always call this at the end of the if/else tree? That would remove all the `break` statements as well.